src/main/kotlin/org/rust/ide/formatter/impl/alignment.kt

```diff
@@ -0,0 +1,32 @@
+package org.rust.ide.formatter.impl
+
+import com.intellij.formatting.Alignment
+import com.intellij.openapi.util.Key
+import com.intellij.psi.tree.TokenSet
+import org.rust.ide.formatter.RustAlignmentStrategy
+import org.rust.ide.formatter.blocks.RustFmtBlock
+import org.rust.lang.core.psi.RustCompositeElementTypes.*
+
+/**
+ * Stores common alignment object for function declarations's parameters, return type & where clause.
+ */
+val PARAMETERS_ALIGNMENT: Key<Alignment> = Key.create("PARAMETERS_ALIGNMENT")
+
+fun RustFmtBlock.getAlignmentStrategy(): RustAlignmentStrategy = when (node.elementType) {
+    TUPLE_EXPR, TUPLE_TYPE, ARG_LIST, ENUM_TUPLE_ARGS ->
+        RustAlignmentStrategy.wrapCond { c, p -> !c.isBlockDelim(p) }
+
+    in PARAMS_LIKE ->
+        RustAlignmentStrategy.lazyWrapCond(
+            { this.getUserData(PARAMETERS_ALIGNMENT) },
+            { c, p -> !c.isBlockDelim(p) })
+
+    in FN_DECLS ->
+        RustAlignmentStrategy.lazyWrapFiltered(
+            { this.getUserData(PARAMETERS_ALIGNMENT) },
+            TokenSet.create(RET_TYPE, WHERE_CLAUSE))
```

Comment:
I would use something like 

```
RunstAlignmentStrategy.inheretedFiltered(
  this, TokenSet.create(RET_TYPE, WHERE_CALUSE)
)
```

This lazyness is a bit misleading, because it is in fact recomputed every time, and may cause strange mutable state related bugs. "Inheritance" is not better in this regard, but it is at least less general and easier to reason about because of this.